### PR TITLE
Query optimizations

### PR DESCRIPTION
**Pull request by benjaminjkraft**
Thanks to #1516, I got nerd sniped a bit and did some query-optimization.  Should hopefully speed up the dashboard a bit.  Depends on #1502, mostly just to avoid merge conflicts.

### CODE REVIEW

**Comment by jmoldow** [comment]
For code reviewers: https://github.com/learning-unlimited/ESP-Website/compare/7225767...query-nerd-sniped

File: esp/esp/program/models/__init__.py [hunk 1]
```diff
@@ -1306,7 +1314,14 @@ def getLastForProgram(user, program):
         if user.is_anonymous():
             regProfList = RegistrationProfile.objects.none()
         else:
-            regProfList = RegistrationProfile.objects.filter(user__exact=user,program__exact=program).select_related().order_by('-last_ts','-id')[:1]
+            regProfList = (RegistrationProfile.objects
+                           .filter(user__exact=user,program__exact=program)
```
Comment (jmoldow) [inline esp/esp/program/models/__init__.py]: Nitpick: space after comma. And is the ``__exact`` needed?

Comment (benjaminjkraft) [inline esp/esp/program/models/__init__.py]: I doubt it, I just reformatted things.  But I can fix while I'm at it.

File: esp/esp/program/modules/handlers/jsondatamodule.py [hunk 2]
```diff
@@ -735,7 +737,9 @@ def student_max_count(clslist):
         for g in grades:
             grade_classes = classes.filter(status__gte=0, grade_min__lte=g, grade_max__gte=g)
             grade_sections = prog.sections().filter(status__gte=0, parent_class__in=grade_classes)
-            grade_students = filter(lambda x: x.getGrade(prog)==g, students['enrolled'])
+            grade_students = filter(
+                lambda x: x.getGrade(prog, assume_student=True)==g,
```
Comment (jmoldow) [inline esp/esp/program/modules/handlers/jsondatamodule.py]: You'll need to rebase on Anthony's changes again. Looks like there is no longer an ``assume_student`` argument.

Comment (jmoldow) [inline esp/esp/program/modules/handlers/jsondatamodule.py]: Never mind, I just realized that's an argument you introduced, not him.

File: esp/esp/users/models/__init__.py [hunk 3]
```diff
@@ -922,14 +935,29 @@ def getGrade(self, program = None):
...
-                    if program:
-                        grade += program.incrementGrade() # adds 1 if appropriate tag is set; else does nothing
+                    return regProf.student_info.graduation_year
+        return None
+    getYOG.get_or_create_token(('self',))
+    getYOG.depend_on_row(lambda: StudentInfo, lambda info: {'self': info.user})
 
+    @cache_function
+    def getGrade(self, program=None, assume_student=False):
+        # assume_student will save us a database hit if the user is a student,
```
Comment (jmoldow) [inline esp/esp/users/models/__init__.py]: nitpick: docstring instead of comment

File: esp/esp/users/models/__init__.py [hunk 4]
```diff
@@ -911,10 +908,26 @@ def current_schoolyear(program=None):
...
+            # "now" is actually whenever the program ran or will run
+            now = program.dates()[0]
+            schoolyear = ESPUser.current_schoolyear(now)
+        schoolyear += program.incrementGrade() # adds 1 if appropriate tag is set; else does nothing
+        return schoolyear
+    program_schoolyear.__func__.depend_on_row(lambda: Tag, lambda tag: {'program': tag.target})
+
+    @cache_function
+    def getYOG(self, program=None, assume_student=False):
+        # assume_student will save us a database hit if the user is a student,
```
Comment (jmoldow) [inline esp/esp/users/models/__init__.py]: nitpick: docstring instead of comment

File: esp/esp/users/models/__init__.py [hunk 5]
```diff
@@ -922,14 +935,29 @@ def getGrade(self, program = None):
...
+                    return regProf.student_info.graduation_year
+        return None
+    getYOG.get_or_create_token(('self',))
+    getYOG.depend_on_row(lambda: StudentInfo, lambda info: {'self': info.user})
 
+    @cache_function
+    def getGrade(self, program=None, assume_student=False):
+        # assume_student will save us a database hit if the user is a student,
+        # but cost us at least one and possibly several if they're not.  See
+        # ESPUser.getYOG above.
```
Comment (jmoldow) [inline esp/esp/users/models/__init__.py]: nitpick: If you change this to a docstring, don't use "above" or "below" when referring to other functions.

**Comment by jmoldow** [comment]
Looks good to me. Just some nitpicks. And we have to wait until #1502 is merged, and then merge in main again.

**Comment by jmoldow** [comment]
#1502 was merged. So I merged main and addressed my nitpicks. Waiting for tests to run.

**Pull request merged**